### PR TITLE
Assign PR assignment based on work queue

### PR DESCRIPTION
**Pull request by apiraino**
This implements the new way of assigning pull requests reviews (designed in #1753). When this PR is merged, PRs in `rust-lang/rust` will be assigned based on user preferences. These can be queried from Zulip (see [docs](https://forge.rust-lang.org/triagebot/pr-assignment-tracking.html)).

If no preferences are set, PRs will be just assigned (like always).

There are multiple points where PRs are assigned (via comment or using the GH UI), I've tried to cover all cases, took some time to test them all. The amount of changes reflect that it is not so easy to catch all spots where this happens. I wish I could make a smaller patch but I think this is an _all-or-nothing_ situation where all cases should be handled at once.

Also modified the zulip triagebot command `work show` to return the # work queue capacity set by the user. This is helpful in case a PR assignment is rejected. A PR to update the [documentation](https://forge.rust-lang.org/triagebot/pr-assignment-tracking.html) will follow-up after merge of this.

If a PR assignment is rejected we will return a message with some context. Examples:

Failed PR assignment to a user:
screenshot-20240403-104629

Failed self-assignment:
screenshot-20240403-104717

r? @jackh726

### CODE REVIEW

File: src/handlers/assign.rs [hunk 1]
```diff
@@ -693,14 +765,57 @@ fn find_reviewer_from_names(
...
+    if filtered_candidates.is_empty() {
+        return Err(FindReviewerError::AllReviewersFiltered {
+            initial: names.to_vec(),
+            filtered: names.to_vec(),
+        });
+    }
+
+    log::debug!("Filtered list of candidates: {:?}", filtered_candidates);
+
+    Ok(filtered_candidates
```
Comment (jackh726) [inline src/handlers/assign.rs:783]: Quickly skimming on my phone. I'm curious how we currently users on vacation? I don't see any logic here or in the function below (but I may be missing it).

File: src/handlers/assign.rs [hunk 2]
```diff
@@ -7,6 +7,9 @@
 //! * `@rustbot release-assignment`: Removes the commenter's assignment.
 //! * `r? @user`: Assigns to the given user (PRs only).
 //!
+//! Note: this module does not handle review assignments issued from the
```
Comment (jackh726) [inline src/handlers/assign.rs:10]: Do you handle this elsewhere?

File: src/handlers/assign.rs (hunk 1 above)
Comment (apiraino) [reply src/handlers/assign.rs:783]: Users on vacation are listed in the [triagebot file](https://github.com/rust-lang/rust/blob/ca7d34efa94afe271accf2bd3d44152a5bd6fff1/triagebot.toml#L771) and then filtered out [here](https://github.com/rust-lang/triagebot/blob/c6fa165f057475e34d5e968c89c25b8962b42a43/src/handlers/assign.rs#L726-L738), just right before (inside `candidate_reviewers_from_names`). Then we apply a second pass and filter out team members after reading their work queue.

My plan is to deprecate the "on vacation" thing altogether and revert the patch that introduced it (https://togithub.com/rust-lang/triagebot/pull/1712). To clarify: the "on vacation" feature is sound but should be handled in a more fine-grained way

File: src/handlers/assign.rs (hunk 2 above)
Comment (apiraino) [reply src/handlers/assign.rs:10]: yes, exactly [here](https://github.com/rust-lang/triagebot/pull/1786/files#diff-63f676f3f26949c577c742ff956f9c696a8a640529b4a268068ac205508e4c11R76). 

At this time we don't handle PR assignment done using the GH web UI, so what happens is a bit ugly; in that piece of code I handle the webhook _post-assignment_ and when the assignee(s) don't have capacity I _revert_ the assignment and emit a comment.

(ideally we should handle those assignments in a webhook but I felt this a topic for another pr :slightly_smiling_face: )

File: src/handlers/assign.rs [hunk 3]
```diff
@@ -75,6 +79,27 @@ const NON_DEFAULT_BRANCH: &str =
...
+Please choose another assignee or if you *really* want `{username}` to review this pull request, please ask them to increase their assignment limit.
+
+(see [documentation](https://forge.rust-lang.org/triagebot/pr-assignment-tracking.html))";
+
+const NO_REVIEWER_HAS_CAPACITY: &str = "
+Could not find a reviewer with enough capacity to be assigned at this time. This is a problem.
+
+Please contact us on [#t-infra](https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra) on Zulip.
+
+cc: @Mark-Simulacrum";
```
Comment (jackh726) [inline src/handlers/assign.rs:101]: better to cc me (or you)

File: src/handlers/assign.rs [hunk 4]
```diff
@@ -75,6 +79,27 @@ const NON_DEFAULT_BRANCH: &str =
...
+You have insufficient capacity to be assigned the pull request at this time. PR assignment has been reverted.
+
+Please choose another assignee or increase your assignment limit.
+
+(see [documentation](https://forge.rust-lang.org/triagebot/pr-assignment-tracking.html))";
+
+pub const REVIEWER_HAS_NO_CAPACITY: &str = "
+`{username}` has insufficient capacity to be assigned the pull request at this time. PR assignment has been reverted.
+
+Please choose another assignee or if you *really* want `{username}` to review this pull request, please ask them to increase their assignment limit.
```
Comment (jackh726) [inline src/handlers/assign.rs:92]: I think just leaving it at "Please choose another assignee" is fine

File: src/lib.rs [hunk 5]
```diff
@@ -132,17 +132,26 @@ pub struct ReviewPrefs {
...
+    pub max_assigned_prs: Option<i32>,
 }
 
 impl ReviewPrefs {
     fn to_string(&self) -> String {
+        let capacity = if self.max_assigned_prs.is_some() {
+            format!("{}", self.max_assigned_prs.expect("This can't fail"))
+        } else {
+            String::from("Not set (i.e. unlimited)")
+        };
```
Comment (jackh726) [inline src/lib.rs:144]: Why not just
```rust
let capacity = match self.max_assigned_prs {
  Some(max) =>  format!("{}",max),
  None => String::from("Not set (i.e. unlimited)"),
};
```

**Review by jackh726 — COMMENTED** [review]
A couple small comments - looks good though. Importantly, this doesn't yet enable people to change their max PRs, which is good. In theory, nothing should change.

**Comment by apiraino** [comment]
@jackh726 I've applied your suggestions, thanks! As discussed, let's leave this here for a moment until we first merge a patch allowing people to set their own workload. Going to prepare it in short.

**Comment by apiraino** [comment]
#1790 is up for review

**Comment by jackh726** [comment]
I think you may have misunderstood. I would like to merge this *before* we enable allowing people to change their setting.

File: src/handlers/assign.rs [hunk 6]
```diff
@@ -75,6 +79,27 @@ const NON_DEFAULT_BRANCH: &str =
...
+Please choose another assignee.
+
+(see [documentation](https://forge.rust-lang.org/triagebot/pr-assignment-tracking.html))";
+
+const NO_REVIEWER_HAS_CAPACITY: &str = "
+Could not find a reviewer with enough capacity to be assigned at this time. This is a problem.
+
+Please contact us on [#t-infra](https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra) on Zulip.
+
+cc: @**Jack Huey** @**apiraino**";
```
Comment (jackh726) [inline src/handlers/assign.rs:101]: I don't think this is right for github.

Should be `cc @jackh726 @apiraino`

Comment (apiraino) [reply src/handlers/assign.rs:101]: what was I thinking :skull:

**Comment by apiraino** [comment]
rebased in #1877

**Pull request merged**